### PR TITLE
Add sof_dai_get_tdm_slots() API

### PR DESCRIPTION
**Pull request by brentlu**
Machine driver sof_da7219 is using local functions to support max98373 because the slot number is different from the number in maxim-common module. To solve this, we

1. Add an new API sof_dai_get_tdm_slots() to retrieve tdm slot number from topology in runtime.
2. Rewrite max_98373_hw_params() function in maxim-common module to take advantage of the new tdm slot number query API.
3. Remove the local function in sof_da7219 and use the common code in maxim-common to support max98373

### CODE REVIEW

File: sound/soc/intel/boards/sof_maxim_common.c [hunk 1]
```diff
@@ -72,28 +73,61 @@ static struct snd_soc_dai_link_component max_98373_components[] = {
...
-		if (!strcmp(codec_dai->component->name, MAX_98373_DEV0_NAME)) {
-			/* DEV0 tdm slot configuration */
-			ret = snd_soc_dai_set_tdm_slot(codec_dai, 0x03, 3, 8, 32);
-		} else if (!strcmp(codec_dai->component->name, MAX_98373_DEV1_NAME)) {
-			/* DEV1 tdm slot configuration */
-			ret = snd_soc_dai_set_tdm_slot(codec_dai, 0x0C, 3, 8, 32);
+	for_each_rtd_codec_dais(rtd, i, codec_dai) {
+		if (i >= ARRAY_SIZE(max_98373_tdm_mask)) {
+			dev_err(codec_dai->dev, "invalid codec index %d\n", i);
+			return -ENODEV;
```
Comment (bardliao) [inline sound/soc/intel/boards/sof_maxim_common.c]: Maybe "Currently, only %d amps are supported",  ARRAY_SIZE(max_98373_tdm_mask) is better. And it is more like an EINVAL error.

File: sound/soc/sof/ipc3-topology.c [hunk 2]
```diff
@@ -2509,15 +2509,15 @@ static int sof_ipc3_dai_get_clk(struct snd_sof_dev *sdev, struct snd_sof_dai *da
...
+		case SOF_DAI_PARAM_INTEL_SSP_MCLK:
 			return private->dai_config->ssp.mclk_rate;
-		case SOF_DAI_CLK_INTEL_SSP_BCLK:
+		case SOF_DAI_PARAM_INTEL_SSP_BCLK:
 			return private->dai_config->ssp.bclk_rate;
 		default:
 			break;
 		}
-		dev_err(sdev->dev, "fail to get SSP clk %d rate\n", clk_type);
+		dev_err(sdev->dev, "fail to get SSP clk %d rate\n", param_type);
```
Comment (plbossart) [inline sound/soc/sof/ipc3-topology.c]: Invalid SSP param %d

File: sound/soc/sof/ipc4-topology.c [hunk 3]
```diff
@@ -3169,13 +3169,13 @@ static int sof_ipc4_dai_get_clk(struct snd_sof_dev *sdev, struct snd_sof_dai *da
...
-		case SOF_DAI_CLK_INTEL_SSP_MCLK:
+		switch (param_type) {
+		case SOF_DAI_PARAM_INTEL_SSP_MCLK:
 			return le32_to_cpu(hw_config->mclk_rate);
-		case SOF_DAI_CLK_INTEL_SSP_BCLK:
+		case SOF_DAI_PARAM_INTEL_SSP_BCLK:
 			return le32_to_cpu(hw_config->bclk_rate);
 		default:
-			dev_err(sdev->dev, "Invalid clk type for SSP %d\n", clk_type);
+			dev_err(sdev->dev, "Invalid clk type for SSP %d\n", param_type);
```
Comment (plbossart) [inline sound/soc/sof/ipc4-topology.c]: Invalid SSP param %d

File: sound/soc/intel/boards/sof_maxim_common.c [hunk 4]
```diff
@@ -72,28 +73,61 @@ static struct snd_soc_dai_link_component max_98373_components[] = {
...
+
 static int max_98373_hw_params(struct snd_pcm_substream *substream,
 			       struct snd_pcm_hw_params *params)
 {
 	struct snd_soc_pcm_runtime *rtd = snd_soc_substream_to_rtd(substream);
+	struct snd_soc_dai_link *dai_link = rtd->dai_link;
 	struct snd_soc_dai *codec_dai;
-	int ret = 0;
-	int j;
+	int i, tdm_slots, ret = 0;
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_maxim_common.c]: I prefer one line per variable, and certainly initialized variables on a separate line...

File: sound/soc/intel/boards/sof_maxim_common.c [hunk 5]
```diff
@@ -72,28 +73,61 @@ static struct snd_soc_dai_link_component max_98373_components[] = {
...
+
+			dev_dbg(codec_dai->dev, "set tdm slot: tx %d rx %d slots %d width %d\n",
+				max_98373_tdm_mask[i].tx,
+				max_98373_tdm_mask[i].rx,
+				tdm_slots, params_width(params));
+
+			ret = snd_soc_dai_set_tdm_slot(codec_dai,
+						       max_98373_tdm_mask[i].tx,
+						       max_98373_tdm_mask[i].rx,
+						       tdm_slots,
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_maxim_common.c:144]: Not following, what happen if tdm_slots is 4, how would this work with the masks set as 0x3?

Is this intended to always transmit two channels no matter what the number of slots might be?

A comment or explanation would help!

File: sound/soc/intel/boards/sof_maxim_common.c [hunk 6]
```diff
@@ -9,6 +9,7 @@
 #include <sound/soc-acpi.h>
 #include <sound/soc-dai.h>
 #include <sound/soc-dapm.h>
+#include <sound/sof.h>
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_maxim_common.c:12]: commit message typos: becasue eariler

File: sound/soc/intel/boards/sof_da7219.c [hunk 7]
```diff
@@ -178,42 +178,21 @@ static void da7219_codec_exit(struct snd_soc_pcm_runtime *rtd)
 	snd_soc_component_set_jack(component, NULL, NULL);
 }
 
-static int max98373_hw_params(struct snd_pcm_substream *substream,
-			      struct snd_pcm_hw_params *params)
+static int card_late_probe(struct snd_soc_card *card)
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_da7219.c:181]: commit message maxin-common -> maxim-module

**Review by plbossart — CHANGES_REQUESTED** [review]
Couple of typos and explanations needed @brentlu, otherwise looks good to me.

File: sound/soc/intel/boards/sof_maxim_common.c (hunk 5 above)
Comment (brentlu) [reply sound/soc/intel/boards/sof_maxim_common.c:144]: Yes, max98373 takes two channels as input and let user space to choose the output from one of 'ch0', 'ch1', '(ch0+ch1)/2'. Here the 0x3 tells codec driver to select TDM channel 0 and 1 as input channels.

File: sound/soc/intel/boards/sof_maxim_common.c (hunk 1 above)
Comment (brentlu) [reply sound/soc/intel/boards/sof_maxim_common.c]: done

File: sound/soc/sof/ipc3-topology.c (hunk 2 above)
Comment (brentlu) [reply sound/soc/sof/ipc3-topology.c]: fixed

File: sound/soc/sof/ipc4-topology.c (hunk 3 above)
Comment (brentlu) [reply sound/soc/sof/ipc4-topology.c]: fixed

File: sound/soc/intel/boards/sof_maxim_common.c (hunk 4 above)
Comment (brentlu) [reply sound/soc/intel/boards/sof_maxim_common.c]: done

File: sound/soc/intel/boards/sof_maxim_common.c (hunk 6 above)
Comment (brentlu) [reply sound/soc/intel/boards/sof_maxim_common.c:12]: fixed

File: sound/soc/intel/boards/sof_da7219.c (hunk 7 above)
Comment (brentlu) [reply sound/soc/intel/boards/sof_da7219.c:181]: done

File: sound/soc/intel/boards/sof_maxim_common.c (hunk 1 above)
Comment (brentlu) [reply sound/soc/intel/boards/sof_maxim_common.c]: Type of 'ARRAY_SIZE(max_98373_tdm_mask)' is different under 32 and 64 bit system and results in build error when compiler checks the placeholder %d in string...

File: sound/soc/intel/boards/sof_maxim_common.c (hunk 6 above)
Comment (plbossart) [reply sound/soc/intel/boards/sof_maxim_common.c:12]: still typo: eariler

File: sound/soc/intel/boards/sof_maxim_common.c [hunk 8]
```diff
@@ -72,28 +73,71 @@ static struct snd_soc_dai_link_component max_98373_components[] = {
...
+					ret);
+				return ret;
+			}
+			break;
+		default:
+			dev_dbg(codec_dai->dev, "codec is in I2S mode\n");
+			break;
 		}
 	}
+
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_maxim_common.c]: spurious line.

File: sound/soc/intel/boards/sof_maxim_common.c [hunk 9]
```diff
@@ -72,28 +73,71 @@ static struct snd_soc_dai_link_component max_98373_components[] = {
...
+			}
+
+			dev_dbg(codec_dai->dev, "set tdm slot: tx %d rx %d slots %d width %d\n",
+				max_98373_tdm_mask[i].tx,
+				max_98373_tdm_mask[i].rx,
+				tdm_slots, params_width(params));
+
+			ret = snd_soc_dai_set_tdm_slot(codec_dai,
+						       max_98373_tdm_mask[i].tx,
+						       max_98373_tdm_mask[i].rx,
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_maxim_common.c:143]: I would still create a mask of all tx and rx setups, and verify that they are all compatible with the tdm slots.

For example if more than 2 channels are used, and tdm_slot == 2 then something is incorrect.

File: sound/soc/intel/boards/sof_da7219.c [hunk 10]
```diff
@@ -178,42 +178,21 @@ static void da7219_codec_exit(struct snd_soc_pcm_runtime *rtd)
...
+	struct snd_soc_dapm_context *dapm = &card->dapm;
+	int err;
+
+	if (ctx->amp_type == CODEC_MAX98373) {
+		/* Disable Left and Right Spk pin after boot */
+		snd_soc_dapm_disable_pin(dapm, "Left Spk");
+		snd_soc_dapm_disable_pin(dapm, "Right Spk");
+		err = snd_soc_dapm_sync(dapm);
+		if (err < 0)
+			return err;
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_da7219.c:193]: where does this come from? this wasn't done before, not sure why it's part of this patch.

Comment (brentlu) [reply sound/soc/intel/boards/sof_da7219.c:193]: It's done in 94d2d0897474 ("ASoC: Intel: Boards: tgl_max98373: add dai_trigger function") on TGL platform. The sof_da7219.c driver was implemented earlier for JSL platform so it's missed.

File: sound/soc/intel/boards/sof_maxim_common.c (hunk 9 above)
Comment (brentlu) [reply sound/soc/intel/boards/sof_maxim_common.c:143]: Add checks to make sure tx/rx mask is compatible with tdm slot number.

File: sound/soc/intel/boards/sof_maxim_common.c (hunk 8 above)
Comment (brentlu) [reply sound/soc/intel/boards/sof_maxim_common.c]: removed

File: sound/soc/intel/boards/sof_da7219.c (hunk 10 above)
Comment (bardliao) [reply sound/soc/intel/boards/sof_da7219.c:193]: > It's done in [94d2d08](https://github.com/thesofproject/linux/commit/94d2d08974746bd0f469ec5c1e50c613de558bfc) ("ASoC: Intel: Boards: tgl_max98373: add dai_trigger function") on TGL platform. The sof_da7219.c driver was implemented earlier for JSL platform so it's missed.

Please add the missing part with different commit.

Comment (brentlu) [reply sound/soc/intel/boards/sof_da7219.c:193]: done

File: sound/soc/intel/boards/sof_da7219.c [hunk 11]
```diff
@@ -180,6 +180,19 @@ static void da7219_codec_exit(struct snd_soc_pcm_runtime *rtd)
 
 static int card_late_probe(struct snd_soc_card *card)
 {
+	struct sof_card_private *ctx = snd_soc_card_get_drvdata(card);
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_da7219.c:183]: commit message, please fix the checkpatch error on the SHA1 value

Comment (brentlu) [reply sound/soc/intel/boards/sof_da7219.c:183]: fixed

**Review by plbossart — APPROVED** [review]

**Review by bardliao — APPROVED** [review]

**Pull request merged**